draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+                mock_zs[source_count] = z_value
+                mock_ra[source_count] = RAbase
+                mock_dec[source_count] = decbase
+                if not self.srcs_at_pixel_centers:
+                    mock_ra[source_count] += ang_size * rtheta[zi][si]
+                    mock_dec[source_count] += ang_size * rphi[zi][si]
+
+                source_count += 1
+
+        # Define arrays to hold full source catalog
+        mock_zs_full = np.empty(self.nsources, dtype=mock_zs.dtype)
+        mock_zerrs_full = np.empty(self.nsources, dtype=mock_zs.dtype)
+        mock_ra_full = np.empty(self.nsources, dtype=mock_ra.dtype)
+        mock_dec_full = np.empty(self.nsources, dtype=mock_dec.dtype)
+
+        # Tuple (not list!) of number of sources in each rank
+        # Note: the counts and displacement arguments of Allgatherv are tuples!
+        nsource_tuple = tuple(self.comm_.allgather(nsource_rank))
+        # Tuple (not list!) of displacements of each rank array in full array
+        dspls = tuple(np.insert(arr=np.cumsum(nsource_tuple)[:-1], obj=0, values=0.0))
+        # Gather redshifts
+        recvbuf = [mock_zs_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_zs, len(mock_zs)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+        # Gather redshift errors
+        recvbuf = [mock_zerrs_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_zerrs, len(mock_zs)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+        # Gather theta
+        recvbuf = [mock_ra_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_ra, len(mock_ra)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+        # Gather phi
+        recvbuf = [mock_dec_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_dec, len(mock_dec)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+
+        # Create catalog container
+        mock_catalog = containers.SpectroscopicCatalog(
+            object_id=np.arange(self.nsources, dtype=np.uint64)
+        )
+
+        # Create position and redshift datasets
+        mock_catalog["position"][:] = np.empty(
+            self.nsources, dtype=[("ra", mock_ra.dtype), ("dec", mock_dec.dtype)]
+        )
+        mock_catalog["redshift"][:] = np.empty(
+            self.nsources, dtype=[("z", mock_zs.dtype), ("z_error", mock_zs.dtype)]
+        )
+        # Assign data to catalog container
+        mock_catalog["position"]["ra"][:] = mock_ra_full
+        mock_catalog["position"]["dec"][:] = mock_dec_full
+        mock_catalog["redshift"]["z"][:] = mock_zs_full
+        mock_catalog["redshift"]["z_error"][:] = mock_zerrs_full
+
+        # If we've created the requested number of mocks, prepare to exit
+        if self._count == self.ncats - 1:
+            self.done = True
+
+        return mock_catalog
+
+    def process_finish(self):
+        """Do nothing when last mock has been created."""
+        return None
+
```

Comment:
Do you understand why this needs to be in there? I would have though omitting this implementation was fine.